Спринт 14 -15

IsaevMilan opened the pull request.

IsaevMilan committed:
Домашка спринт 14

IsaevMilan committed:
Домашка спринт 14.v2

kirillNay commented on app/src/main/java/com/example/myplaylistmaker/MediaAdapter.kt:
```diff
@@ -1,43 +1,43 @@
-    //}
 
+    companion object {
+        private const val CLICK_DEBOUNCE_DELAY = 1000L
```
⏫ В названии константных переменных, лучше указывать название величины. В данном случае `CLICK_DEBOUNCE_DELAY_MS`

kirillNay commented on app/src/main/java/com/example/myplaylistmaker/MediaAdapter.kt:
```diff
@@ -1,43 +1,43 @@
+    private val handler = Handler(Looper.getMainLooper())
 
+
+    private val searchActivityObj = SearchActivity()
```
⏫ В данном случае нам будет правильнее обращаться не к `Activity`, а напрямую к `SearchHistory`. Иначе мы в нашем адаптере создаем Activity - достаточно тяжелый объект - но в дальнейшем его никак не используем (например для открытия)

kirillNay replied:
Вообще лучше стараться всегда себя останавливать, когда есть желание вручную создать `Activity`, ведь этим должна заниматься система

kirillNay commented on app/src/main/java/com/example/myplaylistmaker/MediaPlayerActivity.kt:
```diff
@@ -0,0 +1,161 @@
+
+
+
+    private var mainThreadHandler: Handler? = null
```
⏫`Handler` не привязан к жизненному циклу `Activity`, поэтому мы можем его смело создать до `onCreate` при инициализации `MediaPlayerActivity`

kirillNay commented on app/src/main/java/com/example/myplaylistmaker/MediaPlayerActivity.kt:
```diff
@@ -0,0 +1,161 @@
+    private lateinit var pauseButton: ImageButton
+    private var mediaPlayer = MediaPlayer()
+    lateinit var timer: TextView
+    var time = 0
```
⏫ Поле `time` используется только в классе `MediaPlayerActivity`, поэтому мы можем сделать `private`

kirillNay commented on app/src/main/java/com/example/myplaylistmaker/MediaPlayerActivity.kt:
```diff
@@ -0,0 +1,161 @@
+            "512x512bb.jpg"
+        )
+        if (getImage != "Unknown Cover") {
+            getImage.replace("100x100bb.jpg", "512x512bb.jpg")
```
⏫ `replace("100x100bb.jpg", "512x512bb.jpg")` уже произошел на 64 строчке

kirillNay commented on app/src/main/java/com/example/myplaylistmaker/MediaPlayerActivity.kt:
```diff
@@ -0,0 +1,161 @@
+        mediaPlayer.setOnPreparedListener {
+            playButton.isEnabled = true
+            playerState = STATE_PREPARED
+            playButton.visibility = View.VISIBLE
```
⏫ Вместо использования двух `View` кнопок, мы можем использовать лишь одну и менять их `drawable`. С точки зрения производительности такой подход будет чуть более оптимальный, т.к. мы будет вызывать перерисовку лишь одного элемента вместо двух

kirillNay commented on app/src/main/java/com/example/myplaylistmaker/MediaPlayerActivity.kt:
```diff
@@ -0,0 +1,161 @@
+        mediaPlayer.release()
+    }
+
+    fun timing(): Runnable {
```
⚠️ Метод `timing()` на каждый вызов возвращает новый объект `Runnable`. Таким образом при каждом новом вызове startPlayer количество колбеков в handler'е будет увеличиваться, что негативно будет сказываться на производительности. 
`Runnable` стоит вынести в поле и передавать его mainThreadHandler при вызове `startPlayer()`. Также стоит не забывать очищать Handler при паузе плеера - чтобы мы лишний раз цикл не гоняли, и при смерти экрана - чтобы не столкнуться с проблемами обновления `View`, когда экрана уже нет и утечками памяти.

kirillNay commented on app/src/main/java/com/example/myplaylistmaker/MediaPlayerActivity.kt:
```diff
@@ -0,0 +1,161 @@
+    }
+
+    private fun pausePlayer() {
+        super.onPause()
```
⚠️ Кажется, что вызов метода жизненного цикла `Activity` тут лишний

kirillNay commented on app/src/main/java/com/example/myplaylistmaker/SearchActivity.kt:
```diff
@@ -235,14 +236,29 @@ const val SEARCH_SHARED_PREFS_KEY = "123"
-            if (actionId == EditorInfo.IME_ACTION_DONE) {
-                requestToServer()
-                // ВЫПОЛНЯЙТЕ ПОИСКОВЫЙ ЗАПРОС ЗДЕСЬ
+        val searchRunnable = Runnable { requestToServer() }
```
⏫ Советую тебе немного изменить структуру твоих методов и вынести их из `onCreate`. Вложенные методы редко являются полезными и зачастую они мешают восприятию кода

kirillNay commented on app/src/main/res/drawable/pause_button.xml:
```diff
@@ -0,0 +1,9 @@
+    android:viewportHeight="100">
+  <path
+      android:pathData="M50,8.33C27,8.33 8.33,27 8.33,50C8.33,73 27,91.67 50,91.67C73,91.67 91.67,73 91.67,50C91.67,27 73,8.33 50,8.33ZM45.83,66.67H37.5V33.33H45.83V66.67ZM62.5,66.67H54.17V33.33H62.5V66.67Z"
+      android:fillColor="#1A1B22"/>
```
⏫ Цвета в векторных изображениях также можно выносить в файл ресурсов. Так, можно организовать более гибкую дизайн-систему приложения.
Кроме того, это поможет экономить занимаемое ресурсами место, если нам нужно иметь иконки разных цветов для разных тем приложения. Ведь вместо того, чтобы создавать две иконки и класть в `drawable` и `drawable-night`, мы можем создать одну иконку в `drawable`, а файл ресурса с цветом будет задан в `values` и `values-night`.

IsaevMilan committed:
Домашка спринт 14.v3
 добавлено пауза трека при переходе в фоновый режим
 исправлено время трека
 рефакторинг SearchActivity

IsaevMilan committed:
Домашка спринт 15.

Charmandik commented on .idea/gradle.xml:
Добавь этот тип файлов в гитигнор, их не нужно отслеживать в гите

Charmandik commented on app/src/main/java/com/example/myplaylistmaker/app/App.kt:
```diff
@@ -0,0 +1,27 @@
+            return savedHistory
+        }
+
+        var trackHistoryList = ArrayList<Track>()
```
Данная переменная не используется, стоит ее удалить

Charmandik commented on app/src/main/java/com/example/myplaylistmaker/data/dto/PlayerRepositoryImpl.kt:
```diff
@@ -0,0 +1,79 @@
+
+    override fun play() {
+        mediaPlayer.start()
+        Log.d("PlayerRepositoryImpl", mediaPlayer.isPlaying.toString())
```
Не рекомендую оставлять логи в коде, так код будет выглядеть чище

IsaevMilan replied:
Ok

Charmandik commented on app/src/main/java/com/example/myplaylistmaker/data/network/RetrofitNetworkClient.kt:
```diff
@@ -0,0 +1,31 @@
+                val body = resp.body() ?: Response()
+                return body.apply { resultCode = resp.code() }
+            } else {
+                return Response().apply { resultCode = 400 }
```
Числа, которые просто так вливаются в код, зачастую очень сложно понять и их называют магическими числами. Используй константы, для того, чтобы дать имена своим числам Рекомендую к прочтению статью на эту тему : https://bureau.ru/soviet/20230406/

IsaevMilan replied:
Понял)

Charmandik commented on app/src/main/java/com/example/myplaylistmaker/domain/models/Track.kt:
```diff
@@ -0,0 +1,28 @@
+private const val SECONDS_IN_MINUTE = 60
+
+data class Track(
+    val trackName: String, // Название композиции
```
Старайся не оставлять комментарии, как правило, они долго не удаляются и визуально загрязняют код

IsaevMilan replied:
Ок

Charmandik commented on app/src/main/java/com/example/myplaylistmaker/presentation/ui/activities/MediaPlayerActivity.kt:
```diff
@@ -58,15 +62,19 @@ class MediaPlayerActivity : AppCompatActivity() {
+         */
+
+        val track = intent.extras?.getSerializable(KEY_TRACK_EXTRA) as? Track
+        playerTrackName.text = track?.trackName ?: "Unknown Track"
```
Рекомендуется выносить строки в константы, используя companion object.
Пример правильной реализации:
class MyClass {
    companion object {
        private const val STRING_NAME = "Твоё значение"
    }
}

Такой подход позволит избежать повторения строковых значений в коде и обеспечит более чистую и легко поддерживаемую структуру.

Charmandik commented on app/src/main/java/com/example/myplaylistmaker/presentation/ui/activities/SearchActivity.kt:
```diff
@@ -90,29 +99,25 @@ class SearchActivity : AppCompatActivity() {
+                historyAdapter.setItems(searchHistoryObj.trackHistoryList)
+                searchHistory.isVisible = true
+            } else {
+                searchHistory.isGone = true
```
Для управления видимостью можно использовать свойство isVisible, устанавив значение false

Charmandik commented on app/src/main/java/com/example/myplaylistmaker/presentation/ui/activities/SearchActivity.kt:
```diff
@@ -262,6 +200,61 @@ class SearchActivity : AppCompatActivity() {
 
+    private fun requestToServer() {
+        if (queryInput.text.isNotEmpty()) {
+            mediaRecycler.visibility = GONE
```
Я бы рекомендовал пользоваться одним подходом в проекте для управления видимостью, так код будет легче читаться и выглядеть чище

Charmandik commented on app/src/main/java/com/example/myplaylistmaker/presentation/ui/adapters/MediaAdapter.kt:
```diff
@@ -0,0 +1,65 @@
+        listDiffer.submitList(newItems)
+    }
+
+    private class TrackItemCallback : DiffUtil.ItemCallback<Track>() {
```
Старайся разносить разные классы или интерфейсы по разным файлам. Один класс/интерфейс в один файл.

Charmandik commented on app/src/main/res/layout/track_item.xml:
```diff
@@ -1,93 +1,92 @@
+        app:layout_constraintEnd_toEndOf="parent"
+        app:layout_constraintStart_toStartOf="parent"
+        app:layout_constraintTop_toTopOf="parent"
+        tools:text="Smells Like Teen Spirit Long loooooong teeeeeext " />
```
Чтобы самому было проще понимать, где в верстке текст, который отображается только в превью студии, а где текст, который "зашит", рекомендую текста, используемые в атрибуте tools, оборачивать в квадратные скобки.

Charmandik commented on app/src/main/res/layout/track_item.xml:
```diff
@@ -1,93 +1,92 @@
-        android:src="@drawable/dot"
-        app:tint="?attr/imageButtonStyle" />
+        <ImageView
+            android:id="@+id/dot"
```
Для более простого использования различных классов view рекомендую использовать следующий трюк: К имени переменной добавлять префикс в виде "аббревиатуры" класса, к которой она относится, для TextView - tv, ImageView - iv, RecyclerView - rv и т.д.. Например, у тебя есть RecyclerView для треков, в этом случае будет rvTracks

Charmandik commented on app/src/main/java/com/example/myplaylistmaker/presentation/ui/adapters/MediaAdapter.kt:
```diff
@@ -0,0 +1,65 @@
+) : RecyclerView.Adapter<MediaViewHolder>() {
+
+    companion object {
+        private const val CLICK_THROTTLE_DELAY = 1000L
```
При именовании констант для промежутка времени рекомендую всегда в конце добавлять единицу времени, например _MILLIS, _IN_SECONDS, _HOURS и т.д.

Charmandik commented on app/src/main/java/com/example/myplaylistmaker/app/App.kt:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    companion object {
+        lateinit var savedHistory: SharedPreferences
```
Вместо использования lateinit, я рекомендую тебе использовать подход с добавлением nullability к типу объекта с оператором ?. и сразу проинициализировать его значением null. Это решение позволяет избежать ошибок, связанных с попытками доступа к неинициализированным объектам, которые могут привести к непредсказуемому поведению или крашам приложения. Хотя это может добавить некоторую дополнительную работу при обращении к объекту через оператор ?. , но это значительно повысит безопасность приложения, так как ты будешь уверен, что объект всегда либо имеет значение, либо равен null, и не будет вызывать исключительных ситуаций.Можешь еще прочитать статью на эту тему https://habr.com/ru/companies/maxilect/articles/590061
